docs/reference/analysis/tokenfilters/word-delimiter-graph-tokenfilter.asciidoc

```diff
@@ -4,105 +4,376 @@
+* `DIGIT` (Numeric)
+* `LOWER` (Lowercase alphabetical)
+* `SUBWORD_DELIM` (Non-alphanumeric delimiter)
+* `UPPER` (Uppercase alphabetical)
+
+This file path must be absolute or relative to the `config` location, and the
+file must be UTF-8 encoded. Each mapping in the file must be separated by a line
+break.
+--
+
+[[analysis-word-delimiter-graph-tokenfilter-customize]]
+==== Customize
+
+To customize the `word_delimiter_graph` filter, duplicate it to create the basis
+for a new custom token filter. You can modify the filter using its configurable
+parameters.
+
+For example, the following request creates a `word_delimiter_graph`
+filter that uses the following rules:
+
+* Split tokens at non-alphanumeric characters, _except_ the hyphen (`-`)
+  character.
+* Remove leading or trailing delimiters from each token.
+* Do _not_ split tokens at letter case transitions.
+* Do _not_ split tokens at letter-number transitions.
+* Remove the English possessive (`'s`) from the end of each token.
+
+[source,console]
+----
+PUT /my_index
+{
+  "settings": {
+    "analysis": {
+      "analyzer": {
+        "default": {
+          "tokenizer": "whitespace",
+          "filter": [ "my_custom_word_delimiter_graph_filter" ]
+        }
+      },
+      "filter": {
+        "my_custom_word_delimiter_graph_filter": {
+          "type": "word_delimiter_graph",
+          "type_table": [ "- => ALPHA" ],
+          "split_on_case_change": false,
+          "split_on_numerics": false,
+          "stem_english_possessive": true
+        }
+      }
+    }
+  }
+}
+----
+
+[[analysis-word-delimiter-graph-differences]]
+==== Differences between `word_delimiter` and `word_delimiter_graph`
+
+Both the <<analysis-word-delimiter-tokenfilter,`word_delimiter`>> and
+`word_delimiter_graph` token filters can produce catenated tokens when any of
+the following parameters are `true`:
+
+ * <<word-delimiter-graph-tokenfilter-catenate-all,`catenate_all`>>
+ * <<word-delimiter-graph-tokenfilter-catenate-numbers,`catenate_numbers`>>
+ * <<word-delimiter-graph-tokenfilter-catenate-words,`catenate_words`>>
+
+When adding these new tokens to a stream, the `word_delimiter` filter places
+catenated tokens _after_ the first delimited token. For example, with
+`catenate_words` set to `true`, the `word_delimiter` filter changes [ `the`,
+`wi-fi`, `is`, `enabled`]  to [`the`, `wi`, **`wifi`**, `fi`, `is`, `enabled` ].
+
+This can cause issues for the <<query-dsl-match-query-phrase,`match_phrase`>>
+query and other queries that rely on the sequence of token streams for matching.
+
+The `word_delimiter_graph` filter places catenated tokens _before_ the first
+delimited token. For example, with `catenate_words` set to `true`, the
+`word_delimiter_graph` filter changes [ `the`, `wi-fi`, `is`, `enabled` ] to
+[ `the`, **`wifi`**, `wi`, `fi`, `is`, `enabled` ].
+
+This better preserves the token stream's original sequence and doesn't usually
+interfere with `match_phrase` or similar queries.
```

Comment:
I don't think this is true?  `catenate_X` parameters break phrase searching in general - for example, searching for the exact phrase `the wifi is enabled` won't match against the token stream above because `fi` introduces an extra position, so `is` is indexes as if it were two positions away from `wifi`.  This is a hard problem in lucene - we don't want to start indexing position lengths because that will make phrase queries much slower.

The advantage of the `_graph` variant is that it produces graphs which can be used at query time to generate several queries, so a *query* for `the wi-fi is enabled` will produce two phrase queries, `the wi fi is enabled` and `the wifi is enabled`.  All good if you've indexed the phrase `the wi-fi is enabled`, as the first query will match.  However, searching for `the wifi is enabled` *won't* match - it's all lowercase in the query, so the filter doesn't recognise the need to break it up, and in the index `wifi` is two positions away from `is`.

Breaking up words with hyphens in is tricky because of the possibility that people will try and search for the word without the hyphen; I think these are probably better dealt with via synonyms.  A better usecase for removing punctuation is for things like part numbers, where you only really want phrase searching within a multi-part token, so you use WDGF with a keyword tokenizer.

Reply:
Aha! The query time bit makes sense to me. I can also add some warnings to the catenate parameters so users know they'll break phrase searches. I'll also amend the intro a bit to cover the product/part number use case.